### PR TITLE
separate concerns

### PR DESCRIPTION
**Pull request by elharo**
@suztomo I refactored this to separate printing and formatting problems from finding problems. This removes some code duplication and should make the logic easier to follow.

### CODE REVIEW

**Review by elharo — COMMENTED** [review]
FYI, this one might be easier to follow if you pull it into an IDE so you can read the whole class as refactored. Github's UI makes large scale changes hard to follow in my experience.

File: dependencies/src/main/java/com/google/cloud/tools/opensource/classpath/LinkageCheckerMain.java [hunk 1]
```diff
@@ -150,16 +187,14 @@ public static void main(String[] arguments)
...
-      System.out.println("Wrote the linkage errors as exclusion file: " + outputExclusionFile);
-      
-      // TODO why do we return an empty set here?
-      return ImmutableSet.of();
-    }
-
     return linkageProblems;
   }
+  
+  private static void writeExclusionFile(Path outputExclusionFile,
```
Comment (suztomo) [inline dependencies/src/main/java/com/google/cloud/tools/opensource/classpath/LinkageCheckerMain.java:193]: It's unused. I see `Problem.writeExclusionFile` takes care of this.

File: dependencies/src/main/java/com/google/cloud/tools/opensource/classpath/LinkageCheckerMain.java [hunk 2]
```diff
@@ -60,25 +60,73 @@ public static void main(String[] arguments)
...
-        if (!linkageProblems.isEmpty()) {
-          System.out.println(
-              "For the details of the linkage errors, see "
-                  + "https://github.com/GoogleCloudPlatform/cloud-opensource-java/wiki/Linkage-Checker-Messages");
-          // Throwing an exception is more test-friendly compared with System.exit(1). The latter
-          // abruptly stops test execution.
-          throw new LinkageCheckResultException(linkageProblems.size());
+
+        Path outputExclusionFile = linkageCheckerArguments.getOutputExclusionFile();
+        // All I/O happens here
```
Comment (suztomo) [inline dependencies/src/main/java/com/google/cloud/tools/opensource/classpath/LinkageCheckerMain.java:69]: "All output happens here"? The input happens at `checkJarFiles` and `checkArtifacts`.

File: dependencies/src/main/java/com/google/cloud/tools/opensource/classpath/LinkageCheckerMain.java [hunk 3]
```diff
@@ -60,25 +60,73 @@ public static void main(String[] arguments)
...
+        if (!problems.linkageProblems.isEmpty()) {
+          // TODO really uncertain about this check. Whether to write an exclusion file is
+          // a separate issue from whether to print the linkage problems.
+          if (outputExclusionFile == null) {
```
Comment (suztomo) [inline dependencies/src/main/java/com/google/cloud/tools/opensource/classpath/LinkageCheckerMain.java:73]: Current check looks good. I feel the similarity with `curl -o <filename>`. If we chose to write the output to the file, it does not print the information on console.

File: dependencies/src/main/java/com/google/cloud/tools/opensource/classpath/LinkageCheckerMain.java [hunk 4]
```diff
@@ -60,25 +60,73 @@ public static void main(String[] arguments)
...
+      }
+      System.out.println(
+          "For the details of the linkage errors, see "
+              + "https://github.com/GoogleCloudPlatform/cloud-opensource-java/wiki/Linkage-Checker-Messages");
+    }
+    
+    void writeExclusionFile(Path path) throws IOException, XMLStreamException, TransformerException {
+      ExclusionFiles.write(path, linkageProblems);
+      System.out.println("Wrote the linkage errors as exclusion file: " + path);
+    }
```
Comment (suztomo) [inline dependencies/src/main/java/com/google/cloud/tools/opensource/classpath/LinkageCheckerMain.java:125]: I like this abstraction of getting the result regardless of the output.

**Review by suztomo — COMMENTED** [review]
It's a nice improvement. Thank you.

File: dependencies/src/main/java/com/google/cloud/tools/opensource/classpath/LinkageCheckerMain.java (hunk 2 above)
Comment (elharo) [reply dependencies/src/main/java/com/google/cloud/tools/opensource/classpath/LinkageCheckerMain.java:69]: removed

File: dependencies/src/main/java/com/google/cloud/tools/opensource/classpath/LinkageCheckerMain.java (hunk 1 above)
Comment (elharo) [reply dependencies/src/main/java/com/google/cloud/tools/opensource/classpath/LinkageCheckerMain.java:193]: removed

**Review by suztomo — APPROVED** [review]
